# **Online Retail Transaction Analysis**

## Objectives

* In this notebook, I will perform the ETL (Extract, Transform, Load) process on the Online Retail Transaction dataset. The aim is to clean, prepare, and transform the raw transaction data so that it is suitable for further analysis, visualisation, statistical testing, and predictive modelling.

* I will check the dataset for duplicate records, missing values, incorrect data types, negative or invalid values, and other inconsistencies. Necessary transformations will then be applied to improve data quality and create useful features for later stages of the project.

## Inputs

* The raw Online Retail Transaction dataset

* Python libraries

## Outputs

* A cleaned and transformed version of the dataset,ready for exploratory data analysis, hypothesis testing, visualisation, and machine learning.



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/umarhussain/Documents/Online Retail Project/Online-Retail-Transaction-Analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/umarhussain/Documents/Online Retail Project/Online-Retail-Transaction-Analysis'

# Extract, Transform, Load - ETL

Import Python libraries 

In [5]:
import pandas as pd
import numpy as np

Import raw dataset

In [6]:
# Load the raw dataset into a pandas DataFrame
df = pd.read_csv('Data/Online Retail.csv')

# Display the first five rows to check that the dataset loaded correctly
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


### Preview the Dataset
The first five and last five rows of the dataset are displayed to get an understanding of the data, including column names, values and overall structure of the dataset.

---

In [7]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom


In [8]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680,France


The first and last five rows show that each row represents an individual transaction. This initial inspection helps confirm that the dataset has loaded correctly and provides an overview of the types of information available for further investigation and cleaning.

### Dataset Information
The dataset information is reviewed to understand its structure, column data types and whether any missing values are present before further investigation and cleaning.

In [9]:
# Display the dataset structure, column data types and non-null counts
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   541909 non-null  int64  
 7   Country      541909 non-null  str    
dtypes: float64(1), int64(2), str(5)
memory usage: 33.1 MB


The dataset contains 541,909 rows and 8 columns. The Description column contains 1,454 missing values, while the remaining columns contain no missing values. InvoiceDate is currently stored as a string and will need to be converted to datetime for time-based analysis.

### Missing Values
I will check the dataset for missing values in each column to identify any incomplete records that may require cleaning.

In [12]:
# Count the number of missing values in each column
df.isnull().sum()

InvoiceNo         0
StockCode         0
Description    1454
Quantity          0
InvoiceDate       0
UnitPrice         0
CustomerID        0
Country           0
dtype: int64

The missing-value check identified 1,454 missing values in the Description column. No missing values were identified in the other columns.

### Investigating Missing Product Descriptions

The Description column contains 1,454 missing values. I will examine these values to determine whether they contain useful transaction information.

In [16]:
# Display transactions where the product description is missing
missing_description = df[df['Description'].isna()]

missing_description.head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,15287,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,15287,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,15287,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,15287,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,15287,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,15287,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,15287,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,15287,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,15287,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,15287,United Kingdom


# Section 2


Section 2 content

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
